README.md
# angular-leaflet-directive

[AngularJS](http://angularjs.org/) directive for the Leaflet Javascript
Library. This software aims to easily embed maps managed by leaflet on your
[Leaflet](http://leaflet.cloudmade.com) project.

See some basic examples:

* [Basic example](http://tombatossals.github.io/angular-leaflet-directive/examples/simple-example.html)
* [Custom parameters example](http://tombatossals.github.io/angular-leaflet-directive/examples/custom-parameters-example.html)
* [Markers example](http://tombatossals.github.io/angular-leaflet-directive/examples/markers-example.html)
* [Polyline example](http://tombatossals.github.io/angular-leaflet-directive/examples/path-example.html)
* [Tile changer example](http://tombatossals.github.io/angular-leaflet-directive/examples/tiles-example.html)
* [Legend example](http://tombatossals.github.io/angular-leaflet-directive/examples/legend-example.html)
* [GeoJson example](http://tombatossals.github.io/angular-leaflet-directive/examples/geojson-example.html)
* [Layers example](http://tombatossals.github.io/angular-leaflet-directive/examples/layers-example.html)


To see it in action, go to the main page where you can find more examples and
some documentation:

 * http://tombatossals.github.com/angular-leaflet-directive


## How to use it

You must include the leaflet-directive dependency on your angular module:
```
var app = angular.module("demoapp", ["leaflet-directive"]);
```

After that, you can change the default values of the directive (if you want) on
your angular controller. For example, you can change the tiles source, the
maxzoom on the leaflet map or the polyline path properties.

```javascript
angular.extend($scope, {
    defaults: {
        tileLayer: "http://{s}.tile.opencyclemap.org/cycle/{z}/{x}/{y}.png",
        maxZoom: 14,
        path: {
            weight: 10,
            color: '#800000',
            opacity: 1
        }
    }
});
```

If you want to set the start of the map to a precise position, you can define
the "center" property of the scope (lat, lng, zoom). It will be updated
interacting on the scope and on the leaflet map in two-way binding. Example:
```javascript
angular.extend($scope, {
    center: {
        lat: 51.505,
        lng: -0.09,
        zoom: 8
    }
});

```
Finally, you must include the markup directive on your HTML page, like this:
```html
<leaflet defaults="defaults" center="center" height="480px" width="640px"></leaflet>
```


## How to contribute

You can use grunt/karma to test your code, and grunt/jshint to lint your code.
First, make sure you have npm and grunt-cli installed globally.

```
# Inside the project dir, install the dependencies
$ npm install

# Set the PATH for the binaries of grunt
$ export PATH=$PATH:node_modules/.bin

# JSHINT
$ grunt jshint
Running "jshint:files" (jshint) task
>> 2 files lint free.

Done, without errors.

# KARMA
$ grunt karma
Running "karma:unit" (karma) task
INFO [karma]: Karma server started at http://localhost:9018/
INFO [launcher]: Starting browser PhantomJS
INFO [PhantomJS 1.9 (Linux)]: Connected on socket id y6MlOmYdWFtvS-F83ZHu
PhantomJS 1.9 (Linux): Executed 8 of 8 SUCCESS (0.438 secs / 0.137 secs)


CHANGELOG
angular-leaflet-directive (0.6.1)
* add: add support for google map layers.

angular-leaflet-directive (0.5.1)
* fix: wrap marker click event broadcast in scope apply

angular-leaflet-directive (0.5.0)
* add options for setting doubleClickZoom
* add options for setting callbacks on map events

angular-leaflet-directive (0.4.0)
* broadcast marker click event

angular-leaflet-directive (0.3.3)
* add binding for main marker
* add minZoom option

angular-leaflet-directive (0.3.2)
* add binding for markers' icons

angular-leaflet-directive (0.3.1)
* add width, height attributes
* support for boundingbox binding

angular-leaflet-directive (0.3)
* Add multiple paths binding support. Now path attribute is discarded and paths should be used instead. See path-example.html for more details.


js/app.js
(function (angular) {

    var app = angular.module("MainPage", ['ngRoute', 'leaflet-directive', 'hljs']);
    app.value('$anchorScroll', angular.noop);
    app.config(['$routeProvider', function($routeProvider) {
        $routeProvider.when('/', {
            templateUrl: 'partials/main.html'
        }).when('/getting-started', {
            templateUrl: 'partials/main.html'
        }).when('/how-to-collaborate', {
            templateUrl: 'partials/how-to-collaborate.html'
        }).when('/basic-examples/:example', {
            templateUrl: 'partials/basic-examples.html'
        }).when('/advanced-examples/:example', {
            templateUrl: 'partials/advanced-examples.html'
        });
    }]);

    app.controller("MainController", [ '$scope', '$route', '$routeParams', '$location', function($scope, $route, $routeParams, $location) {
        $scope.$watch(function() { return $location.path(); }, function(value) {
            if (!value) return;
            var section = value.match(/^\/([^\/])*/);
            if (section.length > 0) {
                $scope.activeTab = section[0].replace(/^\//, "");
            }
        });

        $scope.$watch(function() { return $routeParams.example ; }, function(value) {
            $scope.exampleTab = $routeParams.example;
        });

        $scope.go = function(path) {
            $location.path(path);
        };

    }]);

    app.controller("HeaderController", [ '$scope', '$location', function($scope, $location) {
        angular.extend($scope, {
            center: {
                lat: 40.095,
                lng: -3.823,
                zoom: 3
            },
            defaults: {
                scrollWheelZoom: false,
                tileLayer: "http://{s}.tile.opencyclemap.org/cycle/{z}/{x}/{y}.png",
                tileLayerOptions: {
                    opacity: 0.9,
                    detectRetina: true,
                    reuseTiles: true,
                },
                scrollWheelZoom: false
            }
        });

        $scope.$on('leafletDirectiveMap.click', function(event){
            $location.path("/");
        });
    }]);

    app.controller("CenterController", [ '$scope', function($scope) {
        angular.extend($scope, {
            center: {
                lat: 40.095,
                lng: -3.823,
                zoom: 4
            },
            defaults: {
                scrollWheelZoom: false
            }
        });
    }]);

    app.controller("CustomParametersController", [ '$scope', function($scope) {
        angular.extend($scope, {
            london: {
                lat: 51.505,
                lng: -0.09,
                zoom: 8
            },
            defaults: {
                tileLayer: "http://{s}.tile.opencyclemap.org/cycle/{z}/{x}/{y}.png",
                tileLayerOptions: {
                    opacity: 0.9,
                    detectRetina: true,
                    reuseTiles: true,
                },
                scrollWheelZoom: false
            }
        });
    }]);

    app.controller("DraggingMarkersController", [ '$scope', function($scope) {
        angular.extend($scope, {
            center: {
                lat: 40.095,
                lng: -3.823,
                zoom: 6
            },
            defaults: {
                scrollWheelZoom: false
            },
            markers: {
                Madrid: {
                    lat: 40.095,
                    lng: -3.823,
                    message: "This is Madrid. But you can drag me to another position",
                    focus: true,
                    draggable: true
                },
                Barcelona: {
                    lat: 41.38,
                    lng: 2.18,
                    message: "This is Barcelona. You can't drag me",
                    focus: false,
                    draggable: false
                }
            }
        });
    }]);

    app.controller("MainMarkerController", [ '$scope', function($scope) {
        angular.extend($scope, {
            osloCenter: {
                lat: 59.91,
                lng: 10.75,
                zoom: 12
            },
            osloMarker: {
                lat: 59.91,
                lng: 10.75,
                message: "I want to travel here!",
                focus: true,
                draggable: false
            },
            defaults: {
                scrollWheelZoom: false
            }
        });
    }]);

    app.controller("EventsController", [ '$scope', function($scope) {
        angular.extend($scope, {
            center: {
                lat: 52.374004,
                lng: 4.890359,
                zoom: 7
            },
            defaults: {
                scrollWheelZoom: false
            }
        });

        $scope.eventDetected = "No events yet...";

        $scope.$on('leafletDirectiveMap.zoomstart', function(event){
            $scope.eventDetected = "ZoomStart";
        });

        $scope.$on('leafletDirectiveMap.drag', function(event){
            $scope.eventDetected = "Drag";
        });

        $scope.$on('leafletDirectiveMap.click', function(event){
            $scope.eventDetected = "Click";
        });

        $scope.$on('leafletDirectiveMap.mousemove', function(event){
            $scope.eventDetected = "MouseMove";
        });
    }]);

    app.controller("CustomizeMarkersController", [ '$scope', function($scope) {
        var local_icons = {
            leaf_icon: L.icon({
                iconUrl: 'examples/img/leaf-green.png',
                shadowUrl: 'examples/img/leaf-shadow.png',
                 iconSize:     [38, 95], // size of the icon
                shadowSize:   [50, 64], // size of the shadow
                iconAnchor:   [22, 94], // point of the icon which will correspond to marker's location
                shadowAnchor: [4, 62],  // the same for the shadow
                popupAnchor:  [-3, -76] // point from which the popup should open relative to the iconAnchor
            }),
            default_icon: L.icon({
                iconUrl: 'http://cdn.leafletjs.com/leaflet-0.6.4/images/marker-icon.png',
                shadowUrl: 'http://cdn.leafletjs.com/leaflet-0.6.4/images/marker-shadow.png',
                iconSize: [25, 41],
                iconAnchor: [12, 40],
                popupAnchor: [0, -40],
                shadowSize: [41, 41],
                shadowAnchor: [12, 40]
            }),
            div_icon: L.divIcon({
                iconSize: [230, 0],
                html: 'Using <strong>Bold text as an icon</strong>: Lisbon',
                popupAnchor:  [0, 0]
            }),
            object_icon: {
                iconUrl: 'http://leafletjs.com/docs/images/leaf-orange.png',
                shadowUrl: 'http://leafletjs.com/docs/images/leaf-shadow.png',
                iconSize:     [38, 95],
                shadowSize:   [50, 64],
                iconAnchor:   [22, 94],
                shadowAnchor: [4, 62]
            },
        }

        angular.extend($scope, {
            icons: local_icons
        });

        angular.extend($scope, {
            lisbon: {
                lat: 38.716,
                lng: -9.13,
                zoom: 8
            },
            markers: {
                m1: {
                    lat: 38.716,
                    lng: -9.13,
                    message: "I'm a static marker",
                    icon: local_icons.default_icon,
                },
            },
            defaults: {
                scrollWheelZoom: false
            }
        });
    }]);

    app.controller("PathController", [ '$scope', function($scope) {
        angular.extend($scope, {
            center: {
                lat: 53,
                lng: -3,
                zoom: 6
            },
            paths: {
                p1: {
                    color: '#008000',
                    weight: 3,
                    latlngs: [
                        { lat: 51.50, lng: -0.082 },
                        { lat: 51.751, lng: -1.255 },
                        { lat: 53.48, lng: -2.24 },
                    ],
                }
            },
        });
    }]);

    app.controller("LegendController", [ '$scope', function($scope) {
        angular.extend($scope, {
            amsterdam: {
                lat: 52.35,
                lng: 4.91,
                zoom: 12
            },
            legend: {
                position: 'bottomleft',
                colors: [ '#ff0000', '#28c9ff', '#0000ff', '#ecf386' ],
                labels: [ 'National Cycle Route', 'Regional Cycle Route', 'Local Cycle Network', 'Cycleway' ]
            },
            defaults: {
                tileLayer: "http://{s}.tile.opencyclemap.org/cycle/{z}/{x}/{y}.png",
                scrollWheelZoom: false
            }
        });
    }]);

    app.controller("WorldMapController", [ '$scope', '$http', function($scope, $http) {

        $scope.$on("leafletDirectiveMap.geojsonMouseover", function(ev, leafletEvent) {
            countryMouseover(leafletEvent);
        });

        $scope.$on("leafletDirectiveMap.geojsonClick", function(ev, featureSelected, leafletEvent) {
            countryClick(featureSelected, leafletEvent);
        });

        var continentProperties= {
                "009": {
                        name: 'Oceania',
                        colors: [ '#CC0066', '#993366', '#990066', '#CC3399', '#CC6699' ]
                },
                "019": {
                        name: 'America',
                        colors: [ '#006699', '#336666', '#003366', '#3399CC', '#6699CC' ]
                },
                "150": {
                        name: 'Europe',
                        colors: [ '#FF0000', '#CC3333', '#990000', '#FF3333', '#FF6666' ]
                },
                "002": {
                        name: 'Africa',
                        colors: [ '#00CC00', '#339933', '#009900', '#33FF33', '#66FF66' ]
                },
                "142": {
                        name: 'Asia',
                        colors: [ '#FFCC00', '#CC9933', '#999900', '#FFCC33', '#FFCC66' ]
                },
        };

        angular.extend($scope, {
            center: {
                lat: 40.8471,
                lng: 14.0625,
                zoom: 2
            },
            defaults: {
                scrollWheelZoom: false
            },
            legend: {
                colors: [ '#CC0066', '#006699', '#FF0000', '#00CC00', '#FFCC00' ],
                labels: [ 'Oceania', 'America', 'Europe', 'Africa', 'Asia' ]
            }
        });

        function countryClick(country, event) {
            console.log(country.properties.name);
        }

        // Get a country paint color from the continents array of colors
        function getColor(country) {
            if (!country || !country["region-code"]) {
                return "#FFF";
            }

            var colors = continentProperties[country["region-code"]].colors;
            var index = country["alpha-3"].charCodeAt(0) % colors.length ;
            return colors[index];
        }

        function style(feature) {
            return {
                fillColor: getColor($scope.countries[feature.id]),
                weight: 2,
                opacity: 1,
                color: 'white',
                dashArray: '3',
                fillOpacity: 0.7
            };
        }

        // Mouse over function, called from the Leaflet Map Events
        function countryMouseover(leafletEvent) {
            var layer = leafletEvent.target;
            layer.setStyle({
                weight: 2,
                color: '#666',
                fillColor: 'white'
            });
            layer.bringToFront();
        }

        // Get the countries data from a JSON
        $http.get("examples/json/all.json").success(function(data, status) {

            // Put the countries on an associative array
            $scope.countries = {};
            for (var i=0; i< data.length; i++) {
                var country = data[i];
                $scope.countries[country['alpha-3']] = country;
            }

            // Get the countries geojson data from a JSON
            $http.get("examples/json/countries.geo.json").success(function(data, status) {
                angular.extend($scope, {
                    geojson: {
                        data: data,
                        style: style,
                        resetStyleOnMouseout: true
                    }
                });
            });
        });
    }]);
})(window.angular);
